### PR TITLE
[14.0][ADD] sale_promotion_domain_product_discount

### PR DESCRIPTION
**Pull request by ivs-cetmix**

### CODE REVIEW

**Review by francesco-ooops — CHANGES_REQUESTED** [review]
when discount line is applied on website sale, each "Discount apply on" option has a different message for the discount line displayed in cart, but there is no message when the option selected is "On Domain Matching Products",

see video: https://recordit.co/JORG4kcV49

I would add the description "Discount on selected products"

**Comment by francesco-ooops** [comment]
@isserver1 @GabbasovDinar for the rest, lgtm

**Review by francesco-ooops — APPROVED** [review]
Functional review ok!

**Review by elvise — APPROVED** [review]
LGTM!

**Comment by francesco-ooops** [comment]
@OCA/crm-sales-marketing-maintainers review pls?

**Comment by elvise** [comment]
@chienandalu  can you review this PR? Thanks

**Comment by elvise** [comment]
@OCA/crm-sales-marketing-maintainers is there someone for this PR ? 😇

File: sale_promotion_domain_product_discount/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    "summary": "Apply discount only to the domain matching products",
+    "author": "Ooops, Cetmix, Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "category": "Sales Management",
+    "website": "https://github.com/OCA/sale-promotion",
+    "depends": ["sale_coupon"],
+    "external_dependencies": {},
+    "demo": [],
+    "data": [],
+    "qweb": [],
```
Comment (chienandalu) [inline sale_promotion_domain_product_discount/__manifest__.py]:
```suggestion
```

**Comment by francesco-ooops** [comment]
@simahawk can you take a look at this one too? thanks :)

File: sale_promotion_domain_product_discount/README.rst [hunk 2]
```diff
@@ -0,0 +1,35 @@
+**This file is going to be generated by oca-gen-addon-readme.**
```
Comment (chienandalu) [inline sale_promotion_domain_product_discount/README.rst]: It's nice (not mandatory anyway) to run oca-gen-addon-readme

File: sale_promotion_domain_product_discount/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (chienandalu) [inline sale_promotion_domain_product_discount/__manifest__.py]: It's a common guideline to add copyright in the relevant code files. Something like:

```suggestion
# Copyright 2022 Cemtix - Ivan Sokolov
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
{
```

**Review by chienandalu — COMMENTED** [review]
Thanks @isserver1 for this module.

Functional review :+1: 

Some formal comments:

File: sale_promotion_domain_product_discount/tests/test_domain_product_discount.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
+from odoo.tests import tagged
+from odoo.tests.common import SavepointCase
+
+
+@tagged("post_install", "-at_install")
+class CetmixTestSaleCouponProgram(SavepointCase):
+    def setUp(self):
+        super(CetmixTestSaleCouponProgram, self).setUp()
```
Comment (chienandalu) [inline sale_promotion_domain_product_discount/tests/test_domain_product_discount.py]:
```suggestion
    def setUpClass(cls):
        super().setUpClass()
```

File: sale_promotion_domain_product_discount/tests/test_domain_product_discount.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        self.largeCabinet = self.env["product.product"].create(
```
Comment (chienandalu) [inline sale_promotion_domain_product_discount/tests/test_domain_product_discount.py]: `SavepointCase` uses the `cls` object, so you should change your case declaration like:

```suggestion
        cls.largeCabinet = cls.env["product.product"].create(
```

**Review by chienandalu — CHANGES_REQUESTED** [review]

**Comment by chienandalu** [comment]
Also, the guidelines recommend to name the modules with the root of the module they're extending (if any). For example, for this case, `sale_coupon_domain_product_discount` would be more appropiate

File: sale_promotion_domain_product_discount/README.rst (hunk 2 above)
Comment (simahawk) [reply sale_promotion_domain_product_discount/README.rst]: IMO it's easier to review the fragments, personally I always write "wait for the bot ;)" here.
In PR the docs can change, it's easier not to lose time on reviews if you don't see anything here.

File: sale_promotion_domain_product_discount/models/coupon_reward.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
+from odoo import _, fields, models
```
Comment (simahawk) [inline sale_promotion_domain_product_discount/models/coupon_reward.py]: add missing license headers everywhere

File: sale_promotion_domain_product_discount/models/coupon_reward.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+class CouponReward(models.Model):
+    _inherit = "coupon.reward"
+
+    discount_apply_on = fields.Selection(
+        selection_add=[("domain_product", "On Domain Matching Products")],
+    )
+
+    def name_get(self):
+        """
```
Comment (simahawk) [inline sale_promotion_domain_product_discount/models/coupon_reward.py]: I've seen this in many places but to me it sounds weird to override the docstring of a method w/ something that is a comment explaining the override.

File: sale_promotion_domain_product_discount/models/sale_order.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def _get_reward_values_discount(self, program):
+        """
+        Override core method to check and apply new discount apply rule
+        """
+        if program.discount_apply_on == "domain_product":
```
Comment (simahawk) [inline sale_promotion_domain_product_discount/models/sale_order.py]: please avoid such huge methods. Split the logic in smaller, more readable and more test-able methods.

File: sale_promotion_domain_product_discount/models/sale_order.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+                        )
+                        reward_dict[val]["name"] += (
+                            "( " + _("limited to ") + add_name + ")"
+                        )
+                    amount_already_given += reward_dict[val]["price_unit"]
+                    if reward_dict[val]["price_unit"] == 0:
+                        del reward_dict[val]
+            return reward_dict.values()
+        else:
+            return super()._get_reward_values_discount(program)
```
Comment (simahawk) [inline sale_promotion_domain_product_discount/models/sale_order.py]:
```suggestion
        if program.discount_apply_on != "domain_product":
            return super()._get_reward_values_discount(program)
```
Do this on top and you save indentation for the whole method.

File: sale_promotion_domain_product_discount/readme/CONFIGURE.rst [hunk 10]
```diff
@@ -0,0 +1 @@
+No configuration is required to use this module
```
Comment (simahawk) [inline sale_promotion_domain_product_discount/readme/CONFIGURE.rst]: you can trash this file then

File: sale_promotion_domain_product_discount/readme/CREDITS.rst [hunk 11]
```diff
@@ -0,0 +1 @@
+* Odoo Community Association: `Icon <https://github.com/OCA/maintainer-tools/blob/master/template/module/static/description/icon.svg>`_.
```
Comment (simahawk) [inline sale_promotion_domain_product_discount/readme/CREDITS.rst]: this too

File: sale_promotion_domain_product_discount/tests/test_domain_product_discount.py (hunk 4 above)
Comment (simahawk) [reply sale_promotion_domain_product_discount/tests/test_domain_product_discount.py]: and set tracking_disable for the whole suite :pray:

**Review by simahawk — CHANGES_REQUESTED** [review]

File: sale_promotion_domain_product_discount/models/sale_order.py (hunk 8 above)
Comment (ivs-cetmix) [reply sale_promotion_domain_product_discount/models/sale_order.py]: I'm trying to. However this is not clear for me how to do it in this case

Comment (GabbasovDinar) [reply sale_promotion_domain_product_discount/models/sale_order.py]: I have no idea how to refactor it. In any case, the original method from Odoo has more code than here.

**Comment by GabbasovDinar** [comment]
@chienandalu  Any ideas to fix in travis?
```
File "/home/travis/build/OCA/sale-promotion/sale_coupon_order_line_link/models/sale_order.py", line 73, in _link_reward_discount_lines
lines.filtered(lambda x: x.tax_id == tax).write(
UnboundLocalError: local variable 'lines' referenced before assignment
```

**Comment by chienandalu** [comment]
> @chienandalu Any ideas to fix in travis?
> 
> ```
> File "/home/travis/build/OCA/sale-promotion/sale_coupon_order_line_link/models/sale_order.py", line 73, in _link_reward_discount_lines
> lines.filtered(lambda x: x.tax_id == tax).write(
> UnboundLocalError: local variable 'lines' referenced before assignment
> ```

Yes, as you're adding a new discount scope it doesn't enter here: https://github.com/OCA/sale-promotion/blob/14.0/sale_coupon_order_line_link/models/sale_order.py#L58-L65

I could add a hook method to get extended criterias and a default for that lines values

File: sale_coupon_domain_product_discount/models/sale_order.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+                        "product_id": program.discount_line_product_id.id,
+                        "price_unit": -discount_line_amount
+                        if discount_line_amount > 0
+                        else 0,
+                        "product_uom_qty": 1.0,
+                        "product_uom": uom_id,
+                        "is_reward_line": True,
+                        "tax_id": [(4, tax.id, False) for tax in taxes],
+                    }
+                    currently_discounted_amount += discount_line_amount
```
Comment (simahawk) [inline sale_coupon_domain_product_discount/models/sale_order.py]:
```suggestion
            self._handle_whatever_you_are_handling_here_on_a_line(program, line)
     [....]
     def _handle_whatever_you_are_handling_here_on_a_line(program, line, reward_dict, currently_discounted_amount):
            discount_line_amount = min(
                self._get_reward_values_discount_percentage_per_line(program, line),
                amount_total - currently_discounted_amount,
            )
            if discount_line_amount:
                if line.tax_id in reward_dict:
                    reward_dict[line.tax_id]["price_unit"] -= discount_line_amount
                else:
                    taxes = self.fiscal_position_id.map_tax(line.tax_id)
                    uom_id = program.discount_line_product_id.uom_id.id
                    reward_dict[line.tax_id] = {
                        "name": _(
                            "Discount: %(program)s - On product with following "
                            "taxes: %(taxes)s",
                            program=program.name,
                            taxes=", ".join(taxes.mapped("name")),
                        ),
                        "product_id": program.discount_line_product_id.id,
                        "price_unit": -discount_line_amount
                        if discount_line_amount > 0
                        else 0,
                        "product_uom_qty": 1.0,
                        "product_uom": uom_id,
                        "is_reward_line": True,
                        "tax_id": [(4, tax.id, False) for tax in taxes],
                    }
                    currently_discounted_amount += discount_line_amount
               return $whatever, you, need, to, get, back
```

File: sale_coupon_domain_product_discount/models/sale_order.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+                self._get_reward_values_discount_percentage_per_line(program, line),
+                amount_total - currently_discounted_amount,
+            )
+            if discount_line_amount:
+                if line.tax_id in reward_dict:
+                    reward_dict[line.tax_id]["price_unit"] -= discount_line_amount
+                else:
+                    taxes = self.fiscal_position_id.map_tax(line.tax_id)
+                    uom_id = program.discount_line_product_id.uom_id.id
+                    reward_dict[line.tax_id] = {
```
Comment (simahawk) [inline sale_coupon_domain_product_discount/models/sale_order.py]: or move only this to `_get_reward_values_for_tax`

File: sale_coupon_domain_product_discount/models/sale_order.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+                    currently_discounted_amount += discount_line_amount
+
+        # If there is a max amount for discount, we might have to limit
+        # some discount lines or completely remove some lines
+        max_amount = program._compute_program_amount(
+            "discount_max_amount", self.currency_id
+        )
+        if max_amount > 0:
+            amount_already_given = 0
+            for val in list(reward_dict):
```
Comment (simahawk) [inline sale_coupon_domain_product_discount/models/sale_order.py]: you could split this part to `_get_final_amount`

File: sale_promotion_domain_product_discount/models/sale_order.py (hunk 8 above)
Comment (simahawk) [reply sale_promotion_domain_product_discount/models/sale_order.py]: I posted some examples... then you can use your fantasy :wink: 

Is not a blocker but your future selves and everyone else future self will be happy about this in the long term.
Is going to be more readable and easier to add unit tests and/or debug.

**Comment by simahawk** [comment]
> > @chienandalu Any ideas to fix in travis?
> > ```
> > File "/home/travis/build/OCA/sale-promotion/sale_coupon_order_line_link/models/sale_order.py", line 73, in _link_reward_discount_lines
> > lines.filtered(lambda x: x.tax_id == tax).write(
> > UnboundLocalError: local variable 'lines' referenced before assignment
> > ```
> 
> Yes, as you're adding a new discount scope it doesn't enter here: https://github.com/OCA/sale-promotion/blob/14.0/sale_coupon_order_line_link/models/sale_order.py#L58-L65
> 
> I could add a hook method to get extended criterias and a default for that lines values

sounds like the code is just broken here: do you run tests before pushing?

**Comment by chienandalu** [comment]
> > > @chienandalu Any ideas to fix in travis?
> > > ```
> > > File "/home/travis/build/OCA/sale-promotion/sale_coupon_order_line_link/models/sale_order.py", line 73, in _link_reward_discount_lines
> > > lines.filtered(lambda x: x.tax_id == tax).write(
> > > UnboundLocalError: local variable 'lines' referenced before assignment
> > > ```
> > 
> > 
> > Yes, as you're adding a new discount scope it doesn't enter here: https://github.com/OCA/sale-promotion/blob/14.0/sale_coupon_order_line_link/models/sale_order.py#L58-L65
> > I could add a hook method to get extended criterias and a default for that lines values
> 
> sounds like the code is just broken here: do you run tests before pushing?

Well, it's only broken in the context of a non standard discount like this one, that's why it's failing along this new module

**Comment by GabbasovDinar** [comment]
> > @chienandalu Any ideas to fix in travis?
> > ```
> > File "/home/travis/build/OCA/sale-promotion/sale_coupon_order_line_link/models/sale_order.py", line 73, in _link_reward_discount_lines
> > lines.filtered(lambda x: x.tax_id == tax).write(
> > UnboundLocalError: local variable 'lines' referenced before assignment
> > ```
> 
> Yes, as you're adding a new discount scope it doesn't enter here: https://github.com/OCA/sale-promotion/blob/14.0/sale_coupon_order_line_link/models/sale_order.py#L58-L65
> 
> I could add a hook method to get extended criterias and a default for that lines values

Could you do that? Thanks!

**Comment by elvise** [comment]
> > > @chienandalu Any ideas to fix in travis?
> 
> > > ```
> 
> > > File "/home/travis/build/OCA/sale-promotion/sale_coupon_order_line_link/models/sale_order.py", line 73, in _link_reward_discount_lines
> 
> > > lines.filtered(lambda x: x.tax_id == tax).write(
> 
> > > UnboundLocalError: local variable 'lines' referenced before assignment
> 
> > > ```
> 
> > 
> 
> > Yes, as you're adding a new discount scope it doesn't enter here: https://github.com/OCA/sale-promotion/blob/14.0/sale_coupon_order_line_link/models/sale_order.py#L58-L65
> 
> > 
> 
> > I could add a hook method to get extended criterias and a default for that lines values
> 
> 
> 
> Could you do that? Thanks!

@chienandalu @simahawk Can help us for this point ?

**Comment by chienandalu** [comment]
> Can help us for this point ?

Sure, on my way :)

**Comment by chienandalu** [comment]
> > Can help us for this point ?
> 
> Sure, on my way :)

There you are: 

- https://github.com/OCA/sale-promotion/pull/63

**Comment by elvise** [comment]
@GabbasovDinar Merged! :-) 
can you complete this one ?

**Comment by GabbasovDinar** [comment]
@elvise @chienandalu Ready for the review :)

**Comment by elvise** [comment]
@simahawk how does it look to you? :)

File: sale_coupon_domain_product_discount/models/sale_order.py [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+            "price_unit": -discount_line_amount if discount_line_amount > 0 else 0,
+            "product_uom_qty": 1.0,
+            "product_uom": uom_id,
+            "is_reward_line": True,
+            "tax_id": [(4, tax.id, False) for tax in taxes],
+        }
+
+    def _get_final_amount(self, reward_dict, max_amount):
+        amount_already_given = 0
+        for val in list(reward_dict):
```
Comment (simahawk) [inline sale_coupon_domain_product_discount/models/sale_order.py]:
```suggestion
        _reward_dict = reward_dict.copy()
        for tax, val in reward_dict.items():
```
This way you don't alter the original dict, which sounds better for debugging.
Then in the loop alter _reward_dict instead and return it.

File: sale_coupon_domain_product_discount/models/sale_order.py [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+            amount_to_discount = amount_already_given + reward_dict[val]["price_unit"]
+            if abs(amount_to_discount) > max_amount:
+                reward_dict[val]["price_unit"] = -(
+                    max_amount - abs(amount_already_given)
+                )
+                add_name = formatLang(
+                    self.env, max_amount, currency_obj=self.currency_id
+                )
+                reward_dict[val]["name"] += "( " + _("limited to ") + add_name + ")"
```
Comment (simahawk) [inline sale_coupon_domain_product_discount/models/sale_order.py]:
```suggestion
                reward_dict[val]["name"] += _("(limited to %s)", add_name)
```

**Review by simahawk — COMMENTED** [review]
LG, few minor changes missing from my POV

File: sale_coupon_domain_product_discount/models/sale_order.py (hunk 15 above)
Comment (GabbasovDinar) [reply sale_coupon_domain_product_discount/models/sale_order.py]: I'm not sure about the changes:

1. _reward_dict - pre-commit will remove the variable.
2. If we create a variable as 'reward_dict_copy' and we will use in cycle as
```python:
for tax, val in reward_dict_copy.items():
...
    if val["price_unit"] == 0:
        del reward_dict_copy[tax]
```

we will get an error: 

for example:

image

or I didn't understand your idea :)

**Comment by GabbasovDinar** [comment]
@simahawk ready for review

**Comment by elvise** [comment]
@simahawk looks good now ? :)

**Comment by elvise** [comment]
@simahawk sorry if ping you again, but pending only your review :)

**Comment by GabbasovDinar** [comment]
@chienandalu I ported your changes from 13.0

**Review by chienandalu — APPROVED** [review]
Great! Good team up! :smile:

**Comment by pedrobaeza** [comment]
I think you can squash last 3 commits into one. It sounds weird to introduce a diff with 13.0.1.0.0, and again another to 14.0.1.0.0

**Review by francesco-ooops — APPROVED** [review]
Functional review ok!

**Comment by francesco-ooops** [comment]
@simahawk @pedrobaeza this PR is now the porting of #64 , would you say it's ready for merge?

**Review by pedrobaeza — APPROVED** [review]

File: sale_coupon_domain_product_discount/views/coupon_program_views.xml [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+                />
+            </field>
+            <!-- TODO: It'd be better done in fields_view_get -->
+            <xpath
+                expr="//field[@name='discount_specific_product_ids']"
+                position="attributes"
+            >
+                <attribute
+                    name="attrs"
+                >{'invisible': ['|', '|', '|', ('discount_apply_on', '!=', 'specific_products'),('discount_type', '!=', 'percentage'), ('reward_type', 'in', ('product', 'free_shipping')), ('discount_apply_on_domain_product', '!=', False)], 'required': [('reward_type', '=', 'discount'),('discount_apply_on', '=', 'specific_products'),('discount_type', '=', 'percentage'),('discount_apply_on_domain_product', '=', False)]}</attribute>
```
Comment (simahawk) [inline sale_coupon_domain_product_discount/views/coupon_program_views.xml:26]: next time you can define a computed boolean field and use only this to show/hide

**Review by simahawk — APPROVED** [review]

**Pull request merged**